File: app/csharp/tests/Ch06/OptionHandlingTests.cs
using Ch06;
using LanguageExt;
using static LanguageExt.Prelude;
using Xunit;
using static Ch06.OptionHandling;

namespace GrokkingFP.CSharp.Tests.Ch06;

/// <summary>
/// 第6章: Option 型のテスト
/// </summary>
public class OptionHandlingTests
{
    // ============================================
    // SafeDivide のテスト
    // ============================================

    [Fact]
    public void SafeDivide_ReturnsQuotientWhenDivisorIsNotZero()
    {
        Assert.Equal(Some(5), SafeDivide(10, 2));
        Assert.Equal(Some(3), SafeDivide(9, 3));
    }

    [Fact]
    public void SafeDivide_ReturnsNoneWhenDivisorIsZero()
    {
        Assert.Equal(None, SafeDivide(10, 0));
        Assert.Equal(None, SafeDivide(0, 0));
    }

    // ============================================
    // TryParseInt のテスト
    // ============================================

    [Fact]
    public void TryParseInt_ReturnsSomeForValidInteger()
    {
        Assert.Equal(Some(42), TryParseInt("42"));
        Assert.Equal(Some(-10), TryParseInt("-10"));
        Assert.Equal(Some(0), TryParseInt("0"));
    }

    [Fact]
    public void TryParseInt_ReturnsNoneForInvalidInput()
    {
        Assert.Equal(None, TryParseInt("abc"));
        Assert.Equal(None, TryParseInt(""));
        Assert.Equal(None, TryParseInt("12.5"));
    }

    // ============================================
    // TV番組パーサーのテスト
    // ============================================

    [Fact]
    public void ExtractName_ExtractsNameBeforeBracket()
    {
        Assert.Equal(Some("Breaking Bad"), ExtractName("Breaking Bad (2008-2013)"));
        Assert.Equal(Some("Chernobyl"), ExtractName("Chernobyl (2019)"));
    }

    [Fact]
    public void ExtractName_ReturnsNoneForInvalidFormat()
    {
        Assert.Equal(None, ExtractName("No Brackets"));
        Assert.Equal(None, ExtractName("(2019)"));
    }

    [Fact]
    public void ExtractYearStart_ExtractsStartYear()
    {
        Assert.Equal(Some(2008), ExtractYearStart("Breaking Bad (2008-2013)"));
        Assert.Equal(Some(1990), ExtractYearStart("The Simpsons (1990-2023)"));
    }

    [Fact]
    public void ExtractYearStart_ReturnsNoneForSingleYear()
    {
        Assert.Equal(None, ExtractYearStart("Chernobyl (2019)"));
    }

    [Fact]
    public void ExtractYearEnd_ExtractsEndYear()
    {
        Assert.Equal(Some(2013), ExtractYearEnd("Breaking Bad (2008-2013)"));
        Assert.Equal(Some(2023), ExtractYearEnd("The Simpsons (1990-2023)"));
    }

    [Fact]
    public void ExtractYearEnd_ReturnsNoneForSingleYear()
    {
        Assert.Equal(None, ExtractYearEnd("Chernobyl (2019)"));
    }

    [Fact]
    public void ExtractSingleYear_ExtractsSingleYear()
    {
        Assert.Equal(Some(2019), ExtractSingleYear("Chernobyl (2019)"));
        Assert.Equal(Some(2021), ExtractSingleYear("Mare of Easttown (2021)"));
    }

    [Fact]
    public void ExtractSingleYear_ReturnsNoneForYearRange()
    {
        Assert.Equal(None, ExtractSingleYear("Breaking Bad (2008-2013)"));
    }

    [Fact]
    public void ParseShow_ParsesYearRange()
    {
        var result = ParseShow("Breaking Bad (2008-2013)");
        Assert.True(result.IsSome);
        var show = result.Match(Some: s => s, None: () => null!);
        Assert.Equal("Breaking Bad", show.Title);
        Assert.Equal(2008, show.Start);
        Assert.Equal(2013, show.End);
    }

    [Fact]
    public void ParseShow_ParsesSingleYear()
    {
        var result = ParseShow("Chernobyl (2019)");
        Assert.True(result.IsSome);
        var show = result.Match(Some: s => s, None: () => null!);
        Assert.Equal("Chernobyl", show.Title);
        Assert.Equal(2019, show.Start);
        Assert.Equal(2019, show.End);
    }

    [Fact]
    public void ParseShow_ReturnsNoneForInvalidFormat()
    {
        Assert.True(ParseShow("Invalid Format").IsNone);
        Assert.True(ParseShow("No Year ()").IsNone);
    }

    // ============================================
    // 複数番組のパース戦略テスト
    // ============================================

    [Fact]
    public void ParseShowsBestEffort_ReturnsOnlySuccessfulParses()
    {
        var rawShows = Seq(
            "Breaking Bad (2008-2013)",
            "Invalid",
            "Chernobyl (2019)"
        );
        var result = ParseShowsBestEffort(rawShows);
        Assert.Equal(2, result.Count);
    }

    [Fact]
    public void ParseShowsAllOrNothing_ReturnsSomeWhenAllSucceed()
    {
        var rawShows = Seq(
            "Breaking Bad (2008-2013)",
            "Chernobyl (2019)"
        );
        var result = ParseShowsAllOrNothing(rawShows);
        Assert.True(result.IsSome);
        var shows = result.Match(Some: s => s, None: () => Seq<TvShow>());
        Assert.Equal(2, shows.Count);
    }

    [Fact]
    public void ParseShowsAllOrNothing_ReturnsNoneWhenAnyFails()
    {
        var rawShows = Seq(
            "Breaking Bad (2008-2013)",
            "Invalid",
            "Chernobyl (2019)"
        );
        var result = ParseShowsAllOrNothing(rawShows);
        Assert.True(result.IsNone);
    }

    // ============================================
    // AddStrings のテスト
    // ============================================

    [Fact]
    public void AddStrings_AddsTwoValidNumbers()
    {
        Assert.Equal(Some(30), AddStrings("10", "20"));
        Assert.Equal(Some(0), AddStrings("-5", "5"));
    }

    [Fact]
    public void AddStrings_ReturnsNoneWhenEitherIsInvalid()
    {
        Assert.Equal(None, AddStrings("abc", "10"));
        Assert.Equal(None, AddStrings("10", "xyz"));
        Assert.Equal(None, AddStrings("abc", "xyz"));
    }

    // ============================================
    // ForAll/Exists のテスト
    // ============================================

    [Fact]
    public void ForAllExample_ReturnsTrueForNone()
    {
        Assert.True(ForAllExample(None, 10));
    }

    [Fact]
    public void ForAllExample_ReturnsPredicateResultForSome()
    {
        Assert.True(ForAllExample(Some(5), 10));
        Assert.False(ForAllExample(Some(15), 10));
    }

    [Fact]
    public void ExistsExample_ReturnsFalseForNone()
    {
        Assert.False(ExistsExample(None, 10));
    }

    [Fact]
    public void ExistsExample_ReturnsPredicateResultForSome()
    {
        Assert.True(ExistsExample(Some(5), 10));
        Assert.False(ExistsExample(Some(15), 10));
    }
}

/// <summary>
/// ユーザーモデルのテスト
/// </summary>
public class UserModelTests
{
    private readonly Seq<UserModel.User> _users = Seq(
        new UserModel.User("Alice", Some("alice@example.com"), 30),
        new UserModel.User("Bob", None, 25),
        new UserModel.User("Charlie", Some("charlie@other.com"), 35)
    );

    [Fact]
    public void FilterByEmailDomain_IncludesUsersWithNoneEmail()
    {
        var result = UserModel.FilterByEmailDomain(_users, "@example.com");
        // Alice (matches domain) + Bob (None = ForAll returns true)
        Assert.Equal(2, result.Count);
    }

    [Fact]
    public void FilterByEmailDomainExists_ExcludesUsersWithNoneEmail()
    {
        var result = UserModel.FilterByEmailDomainExists(_users, "@example.com");
        // Only Alice (Exists requires Some)
        Assert.Equal(1, result.Count);
        Assert.Equal("Alice", result.ToList().First().Name);
    }
}

File: app/fsharp/tests/Ch06/OptionHandlingTests.fs
module Ch06.OptionHandlingTests

open Xunit
open Ch06.OptionHandling
open Ch06.UserModel

// ============================================
// 安全な除算のテスト
// ============================================

[<Fact>]
let ``safeDivide は正常に除算できる`` () =
    Assert.Equal(Some 5, safeDivide 10 2)
    Assert.Equal(Some 3, safeDivide 7 2)
    Assert.Equal(Some 0, safeDivide 0 5)

[<Fact>]
let ``safeDivide はゼロ除算で None を返す`` () =
    Assert.Equal(None, safeDivide 10 0)
    Assert.Equal(None, safeDivide 0 0)

// ============================================
// 文字列パースのテスト
// ============================================

[<Fact>]
let ``tryParseInt は有効な整数文字列をパースする`` () =
    Assert.Equal(Some 123, tryParseInt "123")
    Assert.Equal(Some -456, tryParseInt "-456")
    Assert.Equal(Some 0, tryParseInt "0")

[<Fact>]
let ``tryParseInt は無効な文字列で None を返す`` () =
    Assert.Equal(None, tryParseInt "abc")
    Assert.Equal(None, tryParseInt "12.34")
    Assert.Equal(None, tryParseInt "")

// ============================================
// TV番組パーサーのテスト
// ============================================

[<Fact>]
let ``extractName は番組名を抽出する`` () =
    Assert.Equal(Some "Breaking Bad", extractName "Breaking Bad (2008-2013)")
    Assert.Equal(Some "The Wire", extractName "The Wire (2002-2008)")

[<Fact>]
let ``extractName は無効な形式で None を返す`` () =
    Assert.Equal(None, extractName "(2008-2013)")
    Assert.Equal(None, extractName "No brackets")

[<Fact>]
let ``extractYearStart は開始年を抽出する`` () =
    Assert.Equal(Some 2008, extractYearStart "Breaking Bad (2008-2013)")
    Assert.Equal(Some 2002, extractYearStart "The Wire (2002-2008)")

[<Fact>]
let ``extractYearStart は無効な形式で None を返す`` () =
    Assert.Equal(None, extractYearStart "Chernobyl (2019)")
    Assert.Equal(None, extractYearStart "Invalid")

[<Fact>]
let ``extractYearEnd は終了年を抽出する`` () =
    Assert.Equal(Some 2013, extractYearEnd "Breaking Bad (2008-2013)")
    Assert.Equal(Some 2008, extractYearEnd "The Wire (2002-2008)")

[<Fact>]
let ``extractSingleYear は単年を抽出する`` () =
    Assert.Equal(Some 2019, extractSingleYear "Chernobyl (2019)")
    Assert.Equal(None, extractSingleYear "Breaking Bad (2008-2013)")

[<Fact>]
let ``parseShow は正常な形式をパースする`` () =
    let result = parseShow "Breaking Bad (2008-2013)"
    Assert.True(result.IsSome)
    let show = result.Value
    Assert.Equal("Breaking Bad", show.Title)
    Assert.Equal(2008, show.Start)
    Assert.Equal(2013, show.End)

[<Fact>]
let ``parseShow は単年の番組をパースする`` () =
    let result = parseShow "Chernobyl (2019)"
    Assert.True(result.IsSome)
    let show = result.Value
    Assert.Equal("Chernobyl", show.Title)
    Assert.Equal(2019, show.Start)
    Assert.Equal(2019, show.End)

[<Fact>]
let ``parseShow は無効な形式で None を返す`` () =
    Assert.True((parseShow "The Wire 2002-2008").IsNone)
    Assert.True((parseShow "(2008-2013)").IsNone)
    Assert.True((parseShow "Invalid").IsNone)

// ============================================
// 複数番組パースのテスト
// ============================================

[<Fact>]
let ``parseShowsBestEffort はパースできたものだけ返す`` () =
    let rawShows = [
        "Breaking Bad (2008-2013)"
        "Invalid Format"
        "Mad Men (2007-2015)"
    ]
    let result = parseShowsBestEffort rawShows

    Assert.Equal(2, List.length result)
    Assert.Equal("Breaking Bad", result.[0].Title)
    Assert.Equal("Mad Men", result.[1].Title)

[<Fact>]
let ``parseShowsAllOrNothing は全部成功したら Some を返す`` () =
    let rawShows = [
        "Breaking Bad (2008-2013)"
        "Mad Men (2007-2015)"
    ]
    let result = parseShowsAllOrNothing rawShows

    Assert.True(result.IsSome)
    Assert.Equal(2, List.length result.Value)

[<Fact>]
let ``parseShowsAllOrNothing は一つでも失敗したら None を返す`` () =
    let rawShows = [
        "Breaking Bad (2008-2013)"
        "Invalid Format"
    ]
    let result = parseShowsAllOrNothing rawShows

    Assert.True(result.IsNone)

// ============================================
// addStrings のテスト
// ============================================

[<Fact>]
let ``addStrings は2つの数値文字列を加算する`` () =
    Assert.Equal(Some 30, addStrings "10" "20")
    Assert.Equal(Some 0, addStrings "5" "-5")

[<Fact>]
let ``addStrings は無効な文字列で None を返す`` () =
    Assert.Equal(None, addStrings "10" "abc")
    Assert.Equal(None, addStrings "xyz" "20")
    Assert.Equal(None, addStrings "abc" "xyz")

// ============================================
// forall と exists のテスト
// ============================================

[<Fact>]
let ``forallExample は Option.forall の動作を示す`` () =
    // Some の場合: 条件をチェック
    Assert.True(forallExample (Some 996) 2020)
    Assert.False(forallExample (Some 996) 500)

    // None の場合: 常に true
    Assert.True(forallExample None 2020)
    Assert.True(forallExample None 500)

[<Fact>]
let ``existsExample は Option.exists の動作を示す`` () =
    // Some の場合: 条件をチェック
    Assert.True(existsExample (Some 996) 2020)
    Assert.False(existsExample (Some 996) 500)

    // None の場合: 常に false
    Assert.False(existsExample None 2020)
    Assert.False(existsExample None 500)

// ============================================
// UserModel のテスト
// ============================================

[<Fact>]
let ``filterByEmailDomain は forall で正しくフィルタする`` () =
    let users = [
        { Name = "Alice"; Email = Some "alice@example.com"; Age = 25 }
        { Name = "Bob"; Email = None; Age = 30 }
        { Name = "Charlie"; Email = Some "charlie@test.com"; Age = 17 }
    ]

    // メールなし or example.com
    let result = filterByEmailDomain "@example.com" users

    Assert.Equal(2, List.length result)
    Assert.Contains(users.[0], result) // Alice
    Assert.Contains(users.[1], result) // Bob (None なので forall = true)

[<Fact>]
let ``filterByEmailDomainExists は exists で正しくフィルタする`` () =
    let users = [
        { Name = "Alice"; Email = Some "alice@example.com"; Age = 25 }
        { Name = "Bob"; Email = None; Age = 30 }
        { Name = "Charlie"; Email = Some "charlie@test.com"; Age = 17 }
    ]

    // メールあり and test.com
    let result = filterByEmailDomainExists "@test.com" users

    Assert.Single(result) |> ignore
    Assert.Equal("Charlie", result.[0].Name)
